cas/ctg_store.c

```diff
@@ -1883,6 +2080,289 @@ static const struct m0_be_btree_kv_ops cas_btree_ops = {
 	.ko_compare = &ctg_cmp
 };
 
+
+static bool ctg_op_is_versioned(const struct m0_ctg_op *ctg_op)
+{
+	const struct m0_cas_op       *cas_op;
+	bool                          has_txd;
+
+	/*
+	 * Since the versioned behavior is optional, it may get turned off
+	 * for a variety of resons. These reasons are listed here
+	 * as M0_RC_INFO messages to aid debugging of the cases
+	 * where the behavior should have not been turned off.
+	 */
+
+	M0_ENTRY("ctg_op=%p", ctg_op);
+
+	if (!ctg_is_ordinary(ctg_op->co_ctg))
+		return M0_RC_INFO(false, "Non-ordinary catalogue.");
+
+	if (ctg_op->co_fom == NULL)
+		return M0_RC_INFO(false, "No fom, no versions.");
+
+	if (ctg_op->co_fom->fo_fop == NULL)
+		return M0_RC_INFO(false, "No fop, no versions.");
+
+	cas_op = m0_fop_data(ctg_op->co_fom->fo_fop);
+	if (cas_op == NULL)
+		return M0_RC_INFO(false, "No cas op, no versions.");
+
+	if ((ctg_op->co_flags & COF_VERSIONED) == 0)
+		return M0_RC_INFO(false, "CAS request is not versioned.");
+
+	has_txd = !m0_dtm0_tx_desc_is_none(&cas_op->cg_txd);
+
+	switch (CTG_OP_COMBINE(ctg_op->co_opcode, ctg_op->co_ct)) {
+	case CTG_OP_COMBINE(CO_PUT, CT_BTREE):
+		if ((ctg_op->co_flags & COF_OVERWRITE) == 0)
+			return M0_RC_INFO(false,
+					  "PUT request without OVERWRITE is "
+					  "not versioned.");
+	case CTG_OP_COMBINE(CO_DEL, CT_BTREE):
+		if (has_txd)
+			return M0_RC(true);
+		else
+			return M0_RC_INFO(false,
+					  "%s request is has an empty txd.",
+					  ctg_op->co_opcode == CO_PUT ?
+					  "PUT" : "DEL");
+	case CTG_OP_COMBINE(CO_GET, CT_BTREE):
+	case CTG_OP_COMBINE(CO_CUR, CT_BTREE):
+		return M0_RC(true);
```

Comment:
All CAS_GET and CAS_NEXT are forced to be versioned operations?

Reply:
Not all. Only GET/NEXT that have COF_VERSIONED flag set. Please take a look at the previous conditions.

Reply:
Yes. OK.